core/src/test/scala/com/evolutiongaming/kafka/flow/RebalanceListenerSpec.scala

```diff
@@ -0,0 +1,125 @@
+        NonEmptySet.of(
+          TopicPartition(topic1, partition1),
+          TopicPartition(topic1, partition2),
+          TopicPartition(topic2, partition3),
+          TopicPartition(topic2, partition4)
+        )
+      )
+      .runS(Context())
+      .value
+
+    val expected = List(
+      Action.Add(topic1, NonEmptySet.of(partition1 -> offset, partition2 -> offset)),
+      Action.Add(topic2, NonEmptySet.of(partition3 -> offset, partition4 -> offset))
+    )
+    assertEquals(result.actions.reverse, expected)
+  }
+
+  test("Listener should handle partitions revoke") {
+    val result = Fixture
+      .listener(topic1, topic2)
+      .onPartitionsRevoked(
+        NonEmptySet.of(
+          TopicPartition(topic1, partition1),
+          TopicPartition(topic2, partition3)
+        )
+      )
+      .runS(Context())
+      .value
+
+    val expected = List(
+      Action.Remove(topic1, NonEmptySet.of(partition1)),
+      Action.Remove(topic2, NonEmptySet.of(partition3))
+    )
+    assertEquals(result.actions.reverse, expected)
+  }
+
+  test("Listener should handle partitions lost") {
+    val result = Fixture
+      .listener(topic1, topic2)
+      .onPartitionsLost(
+        NonEmptySet.of(
+          TopicPartition(topic1, partition1),
+          TopicPartition(topic2, partition3)
+        )
+      )
+      .runS(Context())
+      .value
+
+    val expected = List(
+      Action.Remove(topic1, NonEmptySet.of(partition1)),
+      Action.Remove(topic2, NonEmptySet.of(partition3))
+    )
+    assertEquals(result.actions.reverse, expected)
+  }
+
+}
+
+object RebalanceListenerSpec {
+
+  val topic1 = "topic-1"
+  val topic2 = "topic-2"
+  val partition1 = Partition.unsafe(1)
+  val partition2 = Partition.unsafe(2)
+  val partition3 = Partition.unsafe(3)
+  val partition4 = Partition.unsafe(4)
+  val offset = Offset.unsafe(0)
+
+  type F[A] = State[Context, A]
+
+  implicit val logOf = LogOf.empty[F]
+
+  case class Context(actions: List[Action] = Nil) {
+    def +(action: Action): Context = copy(actions = action :: actions)
+  }
+
+  trait Action
+  object Action {
+    case class Add(topic: String, partitions: NonEmptySet[(Partition, Offset)]) extends Action
+    case class Remove(topic: String, partitions: NonEmptySet[Partition]) extends Action
+    case class Apply(topic: String, records: ConsRecords = ConsRecords.empty) extends Action
+  }
+
+  object Fixture {
+
+    val consumer: Consumer[F] = new Consumer[F] {
+      def subscribe(topics: NonEmptySet[Topic], listener: SRebalanceListener[F]): F[Unit] = ().pure[F]
+      def poll(timeout: FiniteDuration): F[ConsRecords] = ConsRecords.empty.pure[F]
+      def commit(offsets: NonEmptyMap[TopicPartition, OffsetAndMetadata]): F[Unit] = ().pure[F]
+      def position(partition: TopicPartition): F[Offset] = offset.pure[F]
+    }
+
+    def flow(topic: String) = new TopicFlow[F] {
+      def apply(records: ConsRecords): F[Unit] = State modify (_ + Action.Apply(topic, records))
```

Comment:
It does not look `Action.Apply` is used in this test. I would consider removing it and returning `().pure[F]` here.

Reply:
Done